rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+to only have the 'main' worker be pooling for reporting tasks.
+
+#### Telemetry
+
+- Server side fetcher
+  
+The telemetry/server/fetcher.ts will attempt sending the telemetry usage multiple times once per day from each process. 
+We do store a state in the SavedObjects store of the last time the usage was sent to prevent sending multiple times 
+(although race conditions might occur).
+
+- Tasks storing telemetry data
+  
+We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
+tasks will be registered multiple times.
+
+Note that sending the data multiple times doesn’t have any real consequences, so this should be considered 
+non-blocking and only an improvement.
+
+#### TaskManager
+
+Currently, task manager does "claims" for jobs to run based on the server uuid. I think this would still work with 
+workers - each task manager in the worker would be doing "claims" for the same server uuid, which I think is basically
+the same as setting your max_workers to "current max_workers * number of workers"
+
+We probably only want one worker doing the task claims, and then when it gets tasks to run, sends them to a worker to run.
+Short term, having only the 'main' worker perform the task claims and executions should be good enough.
+
+#### Alerting
+
+Do we need the alerting task runner (`TaskRunner`, `TaskRunnerFactory`) to be running on a single worker?
+
+# Summary of the potential user-facing breaking changes
+
+## status API
+
+If we want to have the status API returns each individual worker's status, we will need to change the output of the 
+status API in clustering mode. 
+
+Note that the new format for /api/status is still behind a v8format flag, meaning that if we do these changes before 8.0, 
+we won't be introducing any breaking change later.
+
+## instanceUUID
+
+Depending on our decision regarding the instanceUUID problematic, we may have to change the server.uuid configuration 
+property when clustering mode is enabled, which would be a breaking change.
+
+## distinct logging configuration
+
+If we decide to have each worker output log in distinct files, we would have to change the logging configuration to 
+add a prefix/suffix to each log file, which would be a breaking change. Note that this option is probably not the one 
+we'll choose.
+
+## data folder
+
+We may need to have each worker uses a distinct data folder. Should this be considered a breaking change?
+
+# Drawbacks
+
+- Implementation cost is going to be significant, both in core and in plugins. Also, this will have to be a collaborative
+  effort, as we can't enable the clustered mode in production until all the identified breaking changes have been addressed.
+  
+- Even if easier to deploy, it doesn't really provide anything more than a multi-instances Kibana setup.
```

Comment:
Do we have any side-by-side comparisons of the performance of two modes?

Reply:
> it doesn't really provide anything more than a multi-instances Kibana setup.

Kibana does a lot of background polling work which creates a significant amount of network traffic. If we centralize all this background polling to a single worker we can reduce this background work by a factor equal to the number of worker processes (maybe that would be a factor of 8 or 16 for most deployments?).

It's probably worth trying to quantify the impact, e.g. reducing 1Kb of traffic per month by a factor of 16 is negligible but reducing 10Gb by a factor of 16 would be a big win.

Reply:
> If we centralize all this background polling to a single worker we can reduce this background work by a factor equal to the number of worker processes

That sounds great, but would also require more changes from plugins though.

E.g we could have only one worker perform the license check calls, and then broadcast the result to the other workers, but that means changing the `licensing` plugin implementation.

All these optimization can easily be done later as follow-ups though.

> It's probably worth trying to quantify the impact

We would need to list all the 'background' requests we are performing against ES. Not sure how to do that without help from the other teams though.

Reply:
Agreed, but I think it's worth adding this as a benefit to clustering.

Reply:
This is listed as a potential future benefit of clustering, and the current RFC doesn't preclude us from doing this. However the current proposal is to treat it as an enhancement and not address in the initial implementation.